Src/PerformanceCollector/Filtering/Filter.cs

```diff
@@ -0,0 +1,617 @@
+                   ?? false;
+        }
+
+        private static bool ScanDictionary(IDictionary<string, double> dict, string searchValue)
+        {
+            return dict?.Values.Any(val => val.ToString(CultureInfo.InvariantCulture).IndexOf(searchValue, StringComparison.OrdinalIgnoreCase) != -1)
+                   ?? false;
+        }
+
+        private Expression ProduceComparatorExpressionForSingleFieldCondition(Expression fieldExpression, Type fieldType)
+        {
+            // this must determine an appropriate runtime comparison given the field type, the predicate, and the comparand
+            TypeCode fieldTypeCode = Type.GetTypeCode(fieldType);
+            switch (fieldTypeCode)
+            {
+                case TypeCode.Boolean:
+                    {
+                        this.ThrowOnInvalidFilter(fieldType, !this.comparandBoolean.HasValue);
+
+                        switch (this.predicate)
+                        {
+                            case Predicate.Equal:
+                                Func<bool, bool> comparator = fieldValue => fieldValue == this.comparandBoolean.Value;
+                                return Expression.Call(Expression.Constant(comparator.Target), comparator.Method, fieldExpression);
+                            case Predicate.NotEqual:
+                                comparator = fieldValue => fieldValue != this.comparandBoolean.Value;
+                                return Expression.Call(Expression.Constant(comparator.Target), comparator.Method, fieldExpression);
+                            default:
+                                this.ThrowOnInvalidFilter(fieldType);
+                                break;
+                        }
+                    }
+
+                    break;
+                case TypeCode.Int16:
+                case TypeCode.Int32:
+                case TypeCode.Int64:
+                case TypeCode.UInt16:
+                case TypeCode.UInt32:
+                case TypeCode.UInt64:
+                case TypeCode.Single:
+                case TypeCode.Double:
+                    {
+                        // in order for the expression to compile, we must cast to double unless it's already double
+                        // we're using double as the common lowest denominator for all numerical types
+                        Expression fieldConvertedExpression = fieldTypeCode == TypeCode.Double ? fieldExpression : Expression.ConvertChecked(fieldExpression, typeof(double));
+
+                        switch (this.predicate)
+                        {
+                            case Predicate.Equal:
+                                this.ThrowOnInvalidFilter(fieldType, !this.comparandDouble.HasValue);
+                                return Expression.Equal(fieldConvertedExpression, Expression.Constant(this.comparandDouble.Value));
```

Comment:
Equality in double is somewhat a gamble especially if one cannot guarantee that exactly the same steps were applied to operands on both sides of comparison while they were cast/converted into double. Can it happen here with filters that were set in UI?